### PR TITLE
TapeItem + events

### PR DESCRIPTION
**Pull request by VALERA771**

### CODE REVIEW

**Comment by VALERA771** [comment]
Event are tested and working as intended

File: Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        public TapeItem TapeItem { get; }
+
+        /// <inheritdoc/>
+        public bool IsAllowed { get; set; }
+
+        /// <inheritdoc/>
+        public Player Player { get; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not item has been successfully used.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs:50]:
```suggestion
        /// Gets or sets a value indicating whether the item has been successfully used.
```

File: Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            TapeItem = Item.Get(itemBase).As<TapeItem>();
+            Success = success;
+            IsAllowed = isAllowed;
+        }
+
+        /// <inheritdoc/>
+        public Item Item => TapeItem;
+
+        /// <summary>
+        /// Gets a <see cref="API.Features.Items.TapeItem"/> which is being used.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs:39]:
```suggestion
        /// Gets the <see cref="API.Features.Items.TapeItem"/> which is being used.
```

File: Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace Exiled.Events.EventArgs.Player
+{
+    using Exiled.API.Features;
+    using Exiled.API.Features.Items;
+    using Exiled.Events.EventArgs.Interfaces;
+    using InventorySystem.Items;
+
+    /// <summary>
+    /// Contains all information before player uses a Tape Item.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs:16]:
```suggestion
    /// Contains all information before player uses a tape.
```

File: Exiled.Events/Handlers/Player.cs [hunk 4]
```diff
@@ -517,6 +517,16 @@ public class Player
         /// </summary>
         public static Event<InteractingSnowpileEventArgs> InteractingSnowpile { get; set; } = new();
 
+        /// <summary>
+        /// Invoked before player uses a Tape Item.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:521]:
```suggestion
        /// Invoked before player uses a tape.
```

File: Exiled.Events/Handlers/Player.cs [hunk 5]
```diff
@@ -517,6 +517,16 @@ public class Player
...
+        /// </summary>
+        public static Event<UsingTapeItemEventArgs> UsingTapeItem { get; set; } = new();
+
+        /// <summary>
+        /// Invoked after player uses Tape Item.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:526]:
```suggestion
        /// Invoked after player uses tape.
```

File: Exiled.Events/Handlers/Player.cs [hunk 6]
```diff
@@ -1101,6 +1111,18 @@ public static void OnItemRemoved(ReferenceHub referenceHub, InventorySystem.Item
         /// <param name="ev">The <see cref="InteractingSnowpileEventArgs"/> instance.</param>
         public static void OnInteractingSnowpile(InteractingSnowpileEventArgs ev) => InteractingSnowpile.InvokeSafely(ev);
 
+        /// <summary>
+        /// Called before player uses a Tape Item.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:1115]:
```suggestion
        /// Called before player uses a tape.
```

File: Exiled.Events/Handlers/Player.cs [hunk 7]
```diff
@@ -1101,6 +1111,18 @@ public static void OnItemRemoved(ReferenceHub referenceHub, InventorySystem.Item
...
+        /// </summary>
+        /// <param name="ev">The <see cref="UsingTapeItemEventArgs"/> instance.</param>
+        public static void OnUsingTapeItem(UsingTapeItemEventArgs ev) => UsingTapeItem.InvokeSafely(ev);
+
+        /// <summary>
+        /// Called after player uses Tape Item.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:1121]:
```suggestion
        /// Called after player uses tape.
```

File: Exiled.Events/EventArgs/Player/SuccessfullyUsedTapeItemEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace Exiled.Events.EventArgs.Player
+{
+    using Exiled.API.Features;
+    using Exiled.API.Features.Items;
+    using Exiled.Events.EventArgs.Interfaces;
+    using InventorySystem.Items;
+
+    /// <summary>
+    /// Contains all information after player uses Tape Item.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/SuccessfullyUsedTapeItemEventArgs.cs:16]:
```suggestion
    /// Contains all information after player uses tape.
```

File: Exiled.Events/EventArgs/Player/SuccessfullyUsedTapeItemEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    /// </summary>
+    public class SuccessfullyUsedTapeItemEventArgs : IItemEvent, IPlayerEvent
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SuccessfullyUsedTapeItemEventArgs"/> class.
+        /// </summary>
+        /// <param name="player"><inheritdoc cref="Player"/></param>
+        /// <param name="itemBase"><inheritdoc cref="Item"/></param>
+        public SuccessfullyUsedTapeItemEventArgs(Player player, ItemBase itemBase)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/SuccessfullyUsedTapeItemEventArgs.cs:25]: I'd personally rename it to
```cs
public UsedTapeEventArgs(Player player, ItemBase itemBase)
```

File: Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    public class UsingTapeItemEventArgs : IItemEvent, IDeniableEvent, IPlayerEvent
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="UsingTapeItemEventArgs"/> class.
+        /// </summary>
+        /// <param name="player"><inheritdoc cref="Player"/></param>
+        /// <param name="itemBase"><inheritdoc cref="Item"/></param>
+        /// <param name="success"><inheritdoc cref="Success"/></param>
+        /// <param name="isAllowed"><inheritdoc cref="IsAllowed"/></param>
+        public UsingTapeItemEventArgs(Player player, ItemBase itemBase, bool success, bool isAllowed = true)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/UsingTapeItemEventArgs.cs:27]: I'd personally rename it to
```cs
public UsingTapeEventArgs(Player player, ItemBase itemBase, bool success, bool isAllowed = true)
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Player/SuccessfullyUsedTapeItemEventArgs.cs (hunk 9 above)
Comment (VALERA771) [reply Exiled.Events/EventArgs/Player/SuccessfullyUsedTapeItemEventArgs.cs:25]: Ah yes, forgot a little how this item work

File: Exiled.Events/EventArgs/Player/UsingTapeEventArgs.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        /// <inheritdoc/>
+        public bool IsAllowed { get; set; }
+
+        /// <inheritdoc/>
+        public Player Player { get; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the item has been successfully used.
+        /// </summary>
+        public bool Success { get; set; }
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/UsingTapeEventArgs.cs:52]:
```suggestion
        public bool WasSuccessful { get; set; }
```

File: Exiled.Events/Patches/Events/Player/UsingTapeItem.cs [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+                    // if (!ev.IsAllowed)
+                    //    return;
+                    new(OpCodes.Callvirt, PropertyGetter(typeof(UsingTapeEventArgs), nameof(UsingTapeEventArgs.IsAllowed))),
+                    new(OpCodes.Brfalse_S, retLabel),
+
+                    // success = ev.Success
+                    new(OpCodes.Ldloc_0),
+                    new(OpCodes.Ldloc_S, ev.LocalIndex),
+                    new(OpCodes.Callvirt, PropertyGetter(typeof(UsingTapeEventArgs), nameof(UsingTapeEventArgs.Success))),
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/UsingTapeItem.cs:78]:
```suggestion
                    new(OpCodes.Callvirt, PropertyGetter(typeof(UsingTapeEventArgs), nameof(UsingTapeEventArgs.WasSuccessful))),
```

File: Exiled.Events/Handlers/Player.cs [hunk 13]
```diff
@@ -517,6 +517,16 @@ public class Player
...
 
+        /// <summary>
+        /// Invoked before player uses a tape.
+        /// </summary>
+        public static Event<UsingTapeEventArgs> UsingTapeItem { get; set; } = new();
+
+        /// <summary>
+        /// Invoked after player uses tape.
+        /// </summary>
+        public static Event<UsedTapeItemEventArgs> SuccessfullyUsedTapeItem { get; set; } = new();
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:528]:
```suggestion
        public static Event<UsedTapeItemEventArgs> UsedTape { get; set; } = new();
```

Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:528]:
```suggestion
        public static Event<UsedTapeEventArgs> SuccessfullyUsedTapeItem { get; set; } = new();
```

File: Exiled.Events/Handlers/Player.cs [hunk 14]
```diff
@@ -1101,6 +1111,18 @@ public static void OnItemRemoved(ReferenceHub referenceHub, InventorySystem.Item
...
+        /// Called before player uses a tape.
+        /// </summary>
+        /// <param name="ev">The <see cref="UsingTapeEventArgs"/> instance.</param>
+        public static void OnUsingTapeItem(UsingTapeEventArgs ev) => UsingTapeItem.InvokeSafely(ev);
+
+        /// <summary>
+        /// Called after player uses tape.
+        /// </summary>
+        /// <param name="ev">The <see cref="UsedTapeItemEventArgs"/> instance.</param>
+        public static void OnSuccessfullyUsedTapeItem(UsedTapeItemEventArgs ev) => SuccessfullyUsedTapeItem.InvokeSafely(ev);
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:1124]:
```suggestion
        public static void OnUsedItem(UsedTapeEventArgs ev) => UsedTape.InvokeSafely(ev);
```

File: Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+                    new(OpCodes.Call, Method(typeof(Player), nameof(Player.Get), new[] { typeof(ReferenceHub) })),
+
+                    // this
+                    new(OpCodes.Ldarg_0),
+
+                    // SuccessfullyUsedTapeItemEventArgs ev = new(Player, ItemBase);
+                    new(OpCodes.Newobj, GetDeclaredConstructors(typeof(UsedTapeItemEventArgs))[0]),
+
+                    // Handlers.Player.OnSuccessfullyUsedTapeItem(ev);
+                    new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnSuccessfullyUsedTapeItem))),
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs:52]:
```suggestion
                    new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnUsedTape))),
```

File: Exiled.Events/Handlers/Player.cs [hunk 16]
```diff
@@ -517,6 +517,16 @@ public class Player
         /// </summary>
         public static Event<InteractingSnowpileEventArgs> InteractingSnowpile { get; set; } = new();
 
+        /// <summary>
+        /// Invoked before player uses a tape.
+        /// </summary>
+        public static Event<UsingTapeEventArgs> UsingTapeItem { get; set; } = new();
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:523]:
```suggestion
        public static Event<UsingTapeEventArgs> UsingTape { get; set; } = new();
```

File: Exiled.Events/Handlers/Player.cs [hunk 17]
```diff
@@ -1101,6 +1111,18 @@ public static void OnItemRemoved(ReferenceHub referenceHub, InventorySystem.Item
         /// <param name="ev">The <see cref="InteractingSnowpileEventArgs"/> instance.</param>
         public static void OnInteractingSnowpile(InteractingSnowpileEventArgs ev) => InteractingSnowpile.InvokeSafely(ev);
 
+        /// <summary>
+        /// Called before player uses a tape.
+        /// </summary>
+        /// <param name="ev">The <see cref="UsingTapeEventArgs"/> instance.</param>
+        public static void OnUsingTapeItem(UsingTapeEventArgs ev) => UsingTapeItem.InvokeSafely(ev);
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:1118]:
```suggestion
        public static void OnUsingTape(UsingTapeEventArgs ev) => UsingTapeItem.InvokeSafely(ev);
```

Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:1118]:
```suggestion
        public static void OnUsingTapeItem(UsingTapeEventArgs ev) => UsingTape.InvokeSafely(ev);
```

File: Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+                    // Player.Get(this.Owner);
+                    new(OpCodes.Ldarg_0),
+                    new(OpCodes.Callvirt, PropertyGetter(typeof(InventorySystem.Items.FlamingoTapePlayer.TapeItem), nameof(InventorySystem.Items.FlamingoTapePlayer.TapeItem.Owner))),
+                    new(OpCodes.Call, Method(typeof(Player), nameof(Player.Get), new[] { typeof(ReferenceHub) })),
+
+                    // this
+                    new(OpCodes.Ldarg_0),
+
+                    // SuccessfullyUsedTapeItemEventArgs ev = new(Player, ItemBase);
+                    new(OpCodes.Newobj, GetDeclaredConstructors(typeof(UsedTapeItemEventArgs))[0]),
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs:49]:
```suggestion
                    new(OpCodes.Newobj, GetDeclaredConstructors(typeof(UsedTapeEventArgs))[0]),
```

File: Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+                {
+                    // Player.Get(this.Owner);
+                    new(OpCodes.Ldarg_0),
+                    new(OpCodes.Callvirt, PropertyGetter(typeof(InventorySystem.Items.FlamingoTapePlayer.TapeItem), nameof(InventorySystem.Items.FlamingoTapePlayer.TapeItem.Owner))),
+                    new(OpCodes.Call, Method(typeof(Player), nameof(Player.Get), new[] { typeof(ReferenceHub) })),
+
+                    // this
+                    new(OpCodes.Ldarg_0),
+
+                    // SuccessfullyUsedTapeItemEventArgs ev = new(Player, ItemBase);
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs:48]:
```suggestion
                    // UsedTapeEventArgs ev = new(Player, ItemBase);
```

File: Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+                    new(OpCodes.Newobj, GetDeclaredConstructors(typeof(UsedTapeItemEventArgs))[0]),
+
+                    // Handlers.Player.OnSuccessfullyUsedTapeItem(ev);
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SuccessfullyUsedTapeItem.cs:51]:
```suggestion
                    // Handlers.Player.OnUsedTape(ev);
```

File: Exiled.Events/Patches/Events/Player/UsingTapeItem.cs [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+                    new(OpCodes.Ldc_I4_1),
+
+                    // UsingTapeItemEventArgs ev = new(Player, ItemBase, bool, true);
+                    new(OpCodes.Newobj, GetDeclaredConstructors(typeof(UsingTapeEventArgs))[0]),
+                    new(OpCodes.Dup),
+                    new(OpCodes.Dup),
+                    new(OpCodes.Stloc_S, ev.LocalIndex),
+
+                    // Handlers.Player.OnUsingTapeItem(ev);
+                    new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnUsingTapeItem))),
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/UsingTapeItem.cs:68]:
```suggestion
                    new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnUsingTape))),
```

File: Exiled.Events/Patches/Events/Player/UsingTapeItem.cs [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+                    // true
+                    new(OpCodes.Ldc_I4_1),
+
+                    // UsingTapeItemEventArgs ev = new(Player, ItemBase, bool, true);
+                    new(OpCodes.Newobj, GetDeclaredConstructors(typeof(UsingTapeEventArgs))[0]),
+                    new(OpCodes.Dup),
+                    new(OpCodes.Dup),
+                    new(OpCodes.Stloc_S, ev.LocalIndex),
+
+                    // Handlers.Player.OnUsingTapeItem(ev);
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/UsingTapeItem.cs:67]:
```suggestion
                    // Handlers.Player.OnUsingTape(ev);
```

File: Exiled.Events/Patches/Events/Player/UsingTapeItem.cs [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+                    new(OpCodes.Ldarg_0),
+
+                    // success
+                    new(OpCodes.Ldloc_0),
+                    new(OpCodes.Ldfld, fieldInfo),
+
+                    // true
+                    new(OpCodes.Ldc_I4_1),
+
+                    // UsingTapeItemEventArgs ev = new(Player, ItemBase, bool, true);
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/UsingTapeItem.cs:61]:
```suggestion
                    // UsingTapeEventArgs ev = new(Player, ItemBase, bool, true);
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Player/UsedTapeItemEventArgs.cs [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+{
+    using Exiled.API.Features;
+    using Exiled.API.Features.Items;
+    using Exiled.Events.EventArgs.Interfaces;
+    using InventorySystem.Items;
+
+    /// <summary>
+    /// Contains all information after player uses tape.
+    /// </summary>
+    public class UsedTapeItemEventArgs : IItemEvent, IPlayerEvent
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/UsedTapeItemEventArgs.cs:18]:
```suggestion
    public class UsedTapeEventArgs : IItemEvent, IPlayerEvent
```

File: Exiled.Events/Handlers/Player.cs [hunk 25]
```diff
@@ -1101,6 +1111,18 @@ public static void OnItemRemoved(ReferenceHub referenceHub, InventorySystem.Item
         /// <param name="ev">The <see cref="InteractingSnowpileEventArgs"/> instance.</param>
         public static void OnInteractingSnowpile(InteractingSnowpileEventArgs ev) => InteractingSnowpile.InvokeSafely(ev);
 
+        /// <summary>
+        /// Called before player uses a tape.
+        /// </summary>
+        /// <param name="ev">The <see cref="UsingTapeEventArgs"/> instance.</param>
+        public static void OnUsingTapeItem(UsingTapeEventArgs ev) => UsingTape.InvokeSafely(ev);
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:1118]:
```suggestion
        public static void OnUsingTape(UsingTapeEventArgs ev) => UsingTape.InvokeSafely(ev);
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Comment by louis1706** [comment]
Conflict + this item will certainly be fully rework

**Pull request closed without merging**